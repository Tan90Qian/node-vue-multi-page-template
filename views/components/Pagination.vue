<template>
  <div :class="{ pagination: true, center }">
    <ul class="pagination-list">
      <li
        class="pagination-prev"
        title="上一页"
        :data-disabled="cCurrent === 1"
        @click="handlePageChange('prev');"
      >
        {{ prevText || '&lt;' }}
      </li>
      <template v-if="pageData.length <= 5">
        <li
          :class="['pagination-item', value === cCurrent ? 'actived' : '']"
          v-for="value in pageData"
          :key="value"
          :title="value"
          @click="handlePageChange(value);"
        >
          {{ value }}
        </li>
      </template>
      <template v-else>
        <li
          :class="['pagination-item', cCurrent == 1 ? 'actived' : '']"
          :title="1"
          @click="handlePageChange(1);"
        >
          1
        </li>
        <li v-if="cCurrent >= 5" class="pagination-item" @click="handlePageChange('left');">...</li>
        <li
          :class="['pagination-item', value === cCurrent ? 'actived' : '']"
          v-for="value in showedPage"
          :key="value"
          :title="value"
          @click="handlePageChange(value);"
        >
          {{ value }}
        </li>
        <li
          v-if="cCurrent <= pageData.length - 4"
          class="pagination-item"
          @click="handlePageChange('right');"
        >
          ...
        </li>
        <li
          :class="['pagination-item', cCurrent == pageData.length ? 'actived' : '']"
          :title="pageData.length"
          @click="handlePageChange(pageData.length);"
        >
          {{ pageData.length }}
        </li>
      </template>
      <li
        class="pagination-next"
        title="下一页"
        :data-disabled="cCurrent === pageData.length"
        @click="handlePageChange('next');"
      >
        {{ nextText || '&gt;' }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  props: {
    total: {
      type: Number
    },
    current: {
      type: Number
    },
    pageSize: {
      type: Number
    },
    onChange: {
      type: Function
    },
    prevText: {
      type: String
    },
    nextText: {
      type: String
    },
    center: {
      type: Boolean
    }
  },
  data() {
    return {
      totalNum: 0,
      currentPage: 1,
      size: 10
    };
  },
  computed: {
    cTotal() {
      return this.total || this.totalNum;
    },
    cCurrent() {
      return this.current || this.currentPage;
    },
    cSize() {
      return this.pageSize || this.size;
    },
    pageData() {
      const pageNum = Math.ceil(this.cTotal / this.cSize);
      const result = [1];
      for (let i = 2; i <= pageNum; i++) {
        result.push(i);
      }
      return result;
    },
    showedPage() {
      const { cCurrent, pageData } = this;
      if (cCurrent <= 3) {
        return pageData.slice(1, 5);
      } if (cCurrent >= pageData.length - 2) {
        return pageData.slice(-5, -1);
      }
      return pageData.slice(cCurrent - 3, cCurrent + 2);
    },
    onChangeValid() {
      return this.onChange && typeof this.onChange === 'function';
    }
  },
  methods: {
    handlePageChange(page) {
      const { cCurrent, pageData, onChange, cSize, onChangeValid } = this;
      let target;
      switch (page) {
        case 'prev':
          if (cCurrent === 1) break;
          if (onChangeValid) {
            onChange(cCurrent - 1, cSize);
          } else {
            this.currentPage -= 1;
          }
          break;
        case 'next':
          if (cCurrent === pageData.length) break;
          if (onChangeValid) {
            onChange(cCurrent + 1, cSize);
          } else {
            this.currentPage += 1;
          }
          break;
        case 'left':
          target = cCurrent - 5;
          if (target < 1) target = 1;
          if (onChangeValid) {
            onChange(target, cSize);
          } else {
            this.currentPage = target;
          }
          break;
        case 'right':
          target = cCurrent + 5;
          if (target > pageData.length) target = pageData.length;
          if (onChangeValid) {
            onChange(target, cSize);
          } else {
            this.currentPage = target;
          }
          break;
        default:
          if (onChangeValid) {
            onChange(page, cSize);
          } else {
            this.currentPage = page;
          }
          break;
      }
    }
  }
};
</script>

<style lang="scss">
.pagination-list {
  .pagination-prev,
  .pagination-item,
  .pagination-next {
    float: left;
    box-sizing: border-box;
    min-width: 40px;
    padding: 0 13px;
    border: 1px solid #858f98;
    margin-right: 5px;
    color: #858f98;
    background: #fff;
    font-size: 16px;
    line-height: 38px;
    text-align: center;
    cursor: pointer;
    user-select: none;
  }

  .pagination-next {
    margin-right: 0;
  }

  .pagination-item:hover {
    border-color: #858f98;
  }

  .pagination-item.actived {
    color: #fff;
    background: #061c67;
  }

  .pagination-prev[data-disabled='true'],
  .pagination-next[data-disabled='true'] {
    cursor: not-allowed;
  }

  &::after {
    clear: both;
    display: block;
    content: '';
  }
}
$paginationHeight: 40px;
.pagination.center {
  height: $paginationHeight;
  text-align: center;

  .pagination-list {
    display: inline-block;
    height: $paginationHeight;
  }
}
</style>
