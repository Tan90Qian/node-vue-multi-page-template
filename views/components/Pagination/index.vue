<template>
  <div :class="{ pagination: true, center }">
    <ul class="pagination-list">
      <li
        class="pagination-prev"
        :title="prevText"
        v-if="!(hideOnDisabled && stateCurrent === 1)"
        :data-disabled="stateCurrent === 1"
        @click="handleCurrentChange('prev');"
      >
        {{ prevText }}
      </li>
      <template v-if="pageData.length <= 5">
        <li
          :class="['pagination-item', value === stateCurrent ? 'actived' : '']"
          v-for="value in pageData"
          :key="value"
          :title="value"
          @click="handleCurrentChange(value);"
        >
          {{ value }}
        </li>
      </template>
      <template v-else>
        <li
          :class="['pagination-item', stateCurrent == 1 ? 'actived' : '']"
          :title="1"
          @click="handleCurrentChange(1);"
        >
          1
        </li>
        <li
          v-if="stateCurrent >= 5"
          class="pagination-item left"
          @click="handleCurrentChange('left');"
        >
          ...
        </li>
        <li
          :class="['pagination-item', value === stateCurrent ? 'actived' : '']"
          v-for="value in showedPage"
          :key="value"
          :title="value"
          @click="handleCurrentChange(value);"
        >
          {{ value }}
        </li>
        <li
          v-if="stateCurrent <= pageData.length - 4"
          class="pagination-item right"
          @click="handleCurrentChange('right');"
        >
          ...
        </li>
        <li
          :class="['pagination-item', stateCurrent == pageData.length ? 'actived' : '']"
          :title="pageData.length"
          @click="handleCurrentChange(pageData.length);"
        >
          {{ pageData.length }}
        </li>
      </template>
      <li
        class="pagination-next"
        :title="nextText"
        v-if="!(hideOnDisabled && stateCurrent === pageData.length)"
        :data-disabled="stateCurrent === pageData.length"
        @click="handleCurrentChange('next');"
      >
        {{ nextText }}
      </li>
    </ul>
  </div>
</template>

<script>
import { hasProp } from '../_util/prop-util.js';

function fixControlledValue(value) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export default {
  name: 'Pagination',
  props: {
    total: {
      type: Number,
      default: 0
    },
    defaultCurrent: {
      type: Number,
      default: 1
    },
    current: Number,
    pageSize: {
      type: Number,
      default: 1
    },
    prevText: {
      type: String,
      default: '&lt;'
    },
    nextText: {
      type: String,
      default: '&gt;'
    },
    hideOnDisabled: {
      type: Boolean,
      default: false
    },
    center: {
      type: Boolean,
      default: false
    }
  },
  model: {
    prop: 'current',
    event: 'change'
  },
  data() {
    const { current, defaultCurrent } = this.$props;
    return {
      stateCurrent: fixControlledValue(!hasProp(this, 'current') ? defaultCurrent : current)
    };
  },
  created() {
    if (this.stateCurrent < 1) {
      this.stateCurrent = 1;
    } else if (this.stateCurrent > this.pageData.length) {
      this.stateCurrent = this.pageData.length;
    }
  },
  watch: {
    current(val) {
      this.stateCurrent = fixControlledValue(val);
    }
  },
  computed: {
    pageData() {
      const { total, pageSize } = this;
      const pageNum = Math.ceil(total / pageSize);
      const result = [1];
      for (let i = 2; i <= pageNum; i++) {
        result.push(i);
      }
      return result;
    },
    showedPage() {
      const { stateCurrent, pageData } = this;
      if (stateCurrent <= 3) {
        return pageData.slice(1, 5);
      }
      if (stateCurrent >= pageData.length - 2) {
        return pageData.slice(-5, -1);
      }
      return pageData.slice(stateCurrent - 3, stateCurrent + 2);
    }
  },
  methods: {
    handleCurrentChange(page) {
      const { stateCurrent, pageData, pageSize } = this;
      if (page === stateCurrent) return;
      let target;
      switch (page) {
        case 'prev':
          if (stateCurrent === 1) break;
          this.emitChange(stateCurrent - 1, pageSize);
          break;
        case 'next':
          if (stateCurrent === pageData.length) break;
          this.emitChange(stateCurrent + 1, pageSize);
          break;
        case 'left':
          target = stateCurrent - 5;
          if (target < 1) target = 1;
          this.emitChange(target, pageSize);
          break;
        case 'right':
          target = stateCurrent + 5;
          if (target > pageData.length) target = pageData.length;
          this.emitChange(target, pageSize);
          break;
        default:
          this.emitChange(page, pageSize);
          break;
      }
    },
    emitChange(page, pageSize) {
      if (!hasProp(this, 'current')) {
        this.stateCurrent = page;
      } else {
        this.$emit('change', page, pageSize);
      }
    }
  }
};
</script>

<style lang="scss">
.pagination-list {
  .pagination-prev,
  .pagination-item,
  .pagination-next {
    float: left;
    box-sizing: border-box;
    min-width: 40px;
    padding: 0 13px;
    border: 1px solid #858f98;
    margin-right: 5px;
    color: #858f98;
    background: #fff;
    font-size: 16px;
    line-height: 38px;
    text-align: center;
    cursor: pointer;
    user-select: none;
  }

  .pagination-next {
    margin-right: 0;
  }

  .pagination-item:hover {
    border-color: #858f98;
  }

  .pagination-item.actived {
    color: #fff;
    background: #061c67;
    cursor: default;
  }

  .pagination-prev[data-disabled='true'],
  .pagination-next[data-disabled='true'] {
    cursor: not-allowed;
  }

  &::after {
    clear: both;
    display: block;
    content: '';
  }
}
$paginationHeight: 40px;
.pagination.center {
  height: $paginationHeight;
  text-align: center;

  .pagination-list {
    display: inline-block;
    height: $paginationHeight;
  }
}
</style>
