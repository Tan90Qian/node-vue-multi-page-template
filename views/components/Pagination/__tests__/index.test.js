import { mount } from '@vue/test-utils';

import Pagination from '../index';

describe('Pagination tests', () => {
  /* render test */
  it('should render ul and at least one li', () => {
    const wrapper = mount(Pagination);

    expect(wrapper.contains('.pagination-list')).toBe(true);
    expect(wrapper.contains('li')).toBe(true);
  });

  it('should set stateCurrent with defaultCurrent when current absent', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        defaultCurrent: 10,
        total: 10
      }
    });

    expect(wrapper.vm.stateCurrent).toBe(10);
  });

  it('should compute correct page number', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        pageSize: 3
      }
    });

    expect(wrapper.vm.pageData).toEqual([1, 2, 3, 4]);
  });

  it('initial stateCurrent could not less than 1', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        defaultCurrent: -1
      }
    });

    expect(wrapper.vm.stateCurrent).toBe(1);
  });

  it('initial stateCurrent could not more than page number', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        defaultCurrent: 10,
        total: 10,
        pageSize: 2
      }
    });

    expect(wrapper.vm.stateCurrent).toBe(5);
  });

  it('should add center class on the wrapper', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        center: true
      }
    });

    expect(wrapper.classes()).toContain('center');
  });

  it('should have prev and next item by default', () => {
    const wrapper = mount(Pagination);
    const prev = wrapper.find('.pagination-prev');
    const next = wrapper.find('.pagination-next');

    expect(prev.exists()).toBe(true);
    expect(prev.text()).toBe('&lt;');
    expect(next.exists()).toBe(true);
    expect(next.text()).toBe('&gt;');
  });

  it("should change prev and next item's text", () => {
    const wrapper = mount(Pagination, {
      propsData: {
        prevText: '上一页',
        nextText: '下一页'
      }
    });
    const prev = wrapper.find('.pagination-prev');
    const next = wrapper.find('.pagination-next');

    expect(prev.text()).toBe('上一页');
    expect(next.text()).toBe('下一页');
  });

  it('prev item should have data-disabled attr when stateCurrent equals to 1', () => {
    const wrapper = mount(Pagination);
    const prev = wrapper.find('.pagination-prev');

    expect(prev.attributes('data-disabled')).toBeTruthy();
  });

  it('next item should have data-disabled attr when stateCurrent equals to page number', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 2,
        defaultCurrent: 2
      }
    });
    const next = wrapper.find('.pagination-next');

    expect(next.attributes('data-disabled')).toBeTruthy();
  });

  it('should hide prev item when stateCurrent equals to 1 and hideOnDisabled is true', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        hideOnDisabled: true
      }
    });
    const prev = wrapper.find('.pagination-prev');

    expect(prev.exists()).toBe(false);
  });

  it('should hide next item when stateCurrent equals to page number and hideOnDisabled is true', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        hideOnDisabled: true,
        total: 2,
        defaultCurrent: 2
      }
    });
    const next = wrapper.find('.pagination-next');

    expect(next.exists()).toBe(false);
  });

  it('item which value equals to stateCurrent should have actived class', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        defaultCurrent: 5
      }
    });
    const activedItemArray = wrapper.findAll('.actived');
    const activedItem = activedItemArray.at(0);

    expect(activedItemArray.length).toBe(1);
    expect(activedItem.text()).toBe('5');
  });

  it('should have correct item number when page number less than 6', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 5
      }
    });
    const itemArray = wrapper.findAll('.pagination-item');

    expect(itemArray.length).toBe(5);
  });

  it(`should have correct item number when page number more than 5
      and stateCurrent less than 4`, () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        defaultCurrent: 3
      }
    });
    const itemArray = wrapper.findAll('.pagination-item');
    const left = wrapper.find('.left');
    const right = wrapper.find('.right');

    expect(itemArray.length).toBe(7);
    expect(left.exists()).toBe(false);
    expect(right.exists()).toBe(true);
  });

  it(`should have correct item number when page number more than 5
      and stateCurrent more than (page number - 3)`, () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        defaultCurrent: 8
      }
    });
    const itemArray = wrapper.findAll('.pagination-item');
    const left = wrapper.find('.left');
    const right = wrapper.find('.right');

    expect(itemArray.length).toBe(7);
    expect(left.exists()).toBe(true);
    expect(right.exists()).toBe(false);
  });

  it(`should have correct item number when [page number more than 5]
      and [stateCurrent between 4 and (page number - 3)]`, () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        defaultCurrent: 5
      }
    });
    const itemArray = wrapper.findAll('.pagination-item');
    const left = wrapper.find('.left');
    const right = wrapper.find('.right');

    expect(itemArray.length).toBe(9);
    expect(itemArray.at(2).text()).toBe('3');
    expect(itemArray.at(6).text()).toBe('7');
    expect(left.exists()).toBe(true);
    expect(right.exists()).toBe(true);
  });

  /* interaction test start */

  it('click next will increase current props', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 3,
        current: 1
      },
      listeners: {
        change(page) {
          wrapper.setProps({
            current: page
          });
        }
      }
    });
    const next = wrapper.find('.pagination-next');
    next.trigger('click');

    expect(wrapper.vm.stateCurrent).toBe(2);
  });

  it('click prev will reduce current props', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 3,
        current: 3
      },
      listeners: {
        change(page) {
          wrapper.setProps({
            current: page
          });
        }
      }
    });
    const prev = wrapper.find('.pagination-prev');
    prev.trigger('click');

    expect(wrapper.vm.stateCurrent).toBe(2);
  });

  it('click item will change current to correct value', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 5,
        current: 1
      },
      listeners: {
        change(page) {
          wrapper.setProps({
            current: page
          });
        }
      }
    });
    const target = wrapper.findAll('.pagination-item').at(3);
    target.trigger('click');

    expect(wrapper.vm.stateCurrent).toBe(4);
  });

  it('click actived item will not emit change event', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 5,
        current: 2
      }
    });
    const actived = wrapper.find('.actived');
    actived.trigger('click');

    expect(wrapper.emitted('change')).toBeFalsy();
  });

  it('click left will change current to correct value', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        current: 7
      },
      listeners: {
        change(page) {
          wrapper.setProps({
            current: page
          });
        }
      }
    });
    let left = wrapper.find('.left');
    let actived;

    left.trigger('click');
    actived = wrapper.find('.actived');
    expect(actived.text()).toBe('2');

    wrapper.setProps({ current: 5 });
    left = wrapper.find('.left');
    left.trigger('click');
    actived = wrapper.find('.actived');
    expect(actived.text()).toBe('1');
  });

  it('click right will change current to correct value', () => {
    const wrapper = mount(Pagination, {
      propsData: {
        total: 10,
        current: 2
      },
      listeners: {
        change(page) {
          wrapper.setProps({
            current: page
          });
        }
      }
    });
    let right = wrapper.find('.right');
    let actived;

    right.trigger('click');
    actived = wrapper.find('.actived');
    expect(actived.text()).toBe('7');

    wrapper.setProps({ current: 6 });
    right = wrapper.find('.right');
    right.trigger('click');
    actived = wrapper.find('.actived');
    expect(actived.text()).toBe('10');
  });
});
