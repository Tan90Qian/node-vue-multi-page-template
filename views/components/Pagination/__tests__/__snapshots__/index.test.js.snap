// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination tests renders correctly 1`] = `
<div
  class="pagination"
>
  <ul
    class="pagination-list"
  >
    <li
      class="pagination-prev"
      data-disabled="true"
      title="&lt;"
    >
      &lt;
    </li>
     
    <li
      class="pagination-item actived"
      title="1"
    >
      
        1
      
    </li>
     
    <li
      class="pagination-next"
      data-disabled="true"
      title="&gt;"
    >
      &gt;
    </li>
  </ul>
</div>
`;
